### ISSUE TITLE
[Bug]: [Helm chart] Migration job pod cannot run on OpenShift and behind https proxy

### ISSUE THREAD

**Issue by anetbnd** [BerriAI/litellm#7173]
### What happened?

Hello,

we currently facing issues getting the migration job pod running. We have several issues here:

1. It is missing the proxy environment variables `HTTP_PROXY` and `HTTPS_PROXY` but also `NO_PROXY` (which can be fixed by changing the helm chart accordingly) 

2. On OpenShift the file system is read only by default, therefore when `prisma generate` is called, it tries to install the Prisma CLI and fails, because the standard cache directory is read only (this can be fixed by setting `PRISMA_BINARY_CACHE_DIR` to a temporary directory, but also here a helm chart change is necessary)

3. But the worst problem is, that prisma tries to install nodejs over `python -m nodeenv <cache directory>`. We have an https company proxy, which is not supported by the `urllib.requests` library used by `nodeenv`. So it cannot install nodejs and fails. There are two possible solutions to that:
  * either nodejs is installed inside the Dockerfile for the migration job pod and prisma is setup accordingly to use the preinstalled nodejs version (see https://prisma-client-py.readthedocs.io/en/stable/reference/config/ "Use nodejs-bin")
  * or inside the Dockerfile `RUN prisma version` is executed, which installs the prisma CLI during image build and therefore no further installation is necessary when running the pod on Kubernetes. This would also solve issue 2 and probably also issue 1, depending where the database is running

Is there the chance to add the prisma and nodejs binaries already during the image building?

P.S.: The helm chart version is `0.1.547`.

### Are you a ML Ops Team?

Yes

### What LiteLLM version are you on ?

v1.54.0

**Comment by ghost** [comment]
This should already be fixed with this - https://github.com/BerriAI/litellm/pkgs/container/litellm-non_root

**Comment by freinold** [comment]
We are having the same problem, even when using the non_root image.

Full log of litellm-migrations Job with the latest Helm Chart Version 0.1.625 and latest stable image version "litellm_stable_release_branch-v1.61.13-stable"

```log
[92m14:17:24 - LiteLLM Proxy:ERROR[0m: prisma_migration.py:71 - 'prisma generate' stderr: Error: 
Traceback (most recent call last):
  File "/usr/local/lib/python3.13/site-packages/prisma/generator/generator.py", line 112, in run
    self._on_request(request)
    ~~~~~~~~~~~~~~~~^^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/prisma/generator/generator.py", line 170, in _on_request
    self.generate(data)
    ~~~~~~~~~~~~~^^^^^^
  File "/usr/local/lib/python3.13/site-packages/prisma/generator/generator.py", line 255, in generate
    shutil.copy(data.schema_path, packaged_schema)
    ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/shutil.py", line 428, in copy
    copyfile(src, dst, follow_symlinks=follow_symlinks)
    ~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/shutil.py", line 262, in copyfile
    with open(dst, 'wb') as fdst:
         ~~~~^^^^^^^^^^^
PermissionError: [Errno 13] Permission denied: '/usr/local/lib/python3.13/site-packages/prisma/schema.prisma'



[92m14:17:29 - LiteLLM Proxy:ERROR[0m: prisma_migration.py:83 - 'prisma db push' failed with exit code 1.
[92m14:17:50 - LiteLLM Proxy:ERROR[0m: prisma_migration.py:71 - 'prisma generate' stderr: Error: 
Traceback (most recent call last):
  File "/usr/local/lib/python3.13/site-packages/prisma/generator/generator.py", line 112, in run
    self._on_request(request)
    ~~~~~~~~~~~~~~~~^^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/prisma/generator/generator.py", line 170, in _on_request
    self.generate(data)
    ~~~~~~~~~~~~~^^^^^^
  File "/usr/local/lib/python3.13/site-packages/prisma/generator/generator.py", line 255, in generate
    shutil.copy(data.schema_path, packaged_schema)
    ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/shutil.py", line 428, in copy
    copyfile(src, dst, follow_symlinks=follow_symlinks)
    ~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/shutil.py", line 262, in copyfile
    with open(dst, 'wb') as fdst:
         ~~~~^^^^^^^^^^^
PermissionError: [Errno 13] Permission denied: '/usr/local/lib/python3.13/site-packages/prisma/schema.prisma'



[92m14:17:55 - LiteLLM Proxy:ERROR[0m: prisma_migration.py:83 - 'prisma db push' failed with exit code 1.
[92m14:18:18 - LiteLLM Proxy:ERROR[0m: prisma_migration.py:71 - 'prisma generate' stderr: Error: 
Traceback (most recent call last):
  File "/usr/local/lib/python3.13/site-packages/prisma/generator/generator.py", line 112, in run
    self._on_request(request)
    ~~~~~~~~~~~~~~~~^^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/prisma/generator/generator.py", line 170, in _on_request
    self.generate(data)
    ~~~~~~~~~~~~~^^^^^^
  File "/usr/local/lib/python3.13/site-packages/prisma/generator/generator.py", line 255, in generate
    shutil.copy(data.schema_path, packaged_schema)
    ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/shutil.py", line 428, in copy
    copyfile(src, dst, follow_symlinks=follow_symlinks)
    ~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/shutil.py", line 262, in copyfile
    with open(dst, 'wb') as fdst:
         ~~~~^^^^^^^^^^^
PermissionError: [Errno 13] Permission denied: '/usr/local/lib/python3.13/site-packages/prisma/schema.prisma'



[92m14:18:26 - LiteLLM Proxy:ERROR[0m: prisma_migration.py:83 - 'prisma db push' failed with exit code 1.
[92m14:18:26 - LiteLLM Proxy:ERROR[0m: prisma_migration.py:89 - Unable to push database changes after 3 retries.
```

Seems to me like the non_root image tries to save the generated schema to disk and fails there.

**Comment by ghost** [comment]
@anetbnd can you confirm the non_root image worked for you on v1.61.13?

I haven't seen anyone else report the error @freinold

**Comment by freinold** [comment]
I tried to investigate a littler further: 
It seems like there is already a schema inside the directory, dated to the container build time.

So we probably don't even need to generate the schema again and just have to push it to the DB if we are inside the non_root container.

@krrishdholakia What is the best way to solve this in your opinion?

Is it possible to introduce a new environment variable which gets set during the container build, so it should always be set when running the non-root container?

**Comment by anetbnd** [comment]
@krrishdholakia We did not test fully, but at least the migration job now works without errors when using the non_root image. @mohittalele can give more details.

**Comment by mohittalele** [comment]
@krrishdholakia @anetbnd  I had ran simple test in which it ran successfully with non_root image. As mentioned, we have not tested it fully for e.g - check if data migration itself was properly done. 

I am blocked this week, I would be in position to inform if its work fully next week.

**Comment by freinold** [comment]
@mohittalele would be great if you could provide further info in the coming week.

Especially if you made any changes to the securityContext, annotations or volumes in the values.yaml to get it running inside a non-root cluster.

**Comment by freinold** [comment]
Hey @mohittalele 

checking back in to ask if you could provide more info into how you managed to get non_root running.

Meanwhile on diagnosing the error further:

Our current problem like shown [in the logs above](https://github.com/BerriAI/litellm/issues/7173#issuecomment-2690769643) is the write of the DB schema [yielded by `prisma generate` in the `prisma_migration.py` script](https://github.com/BerriAI/litellm/blob/01c6cbd2705c65119b46866c0b4a44c9d78c662f/litellm/proxy/prisma_migration.py#L63).

My current understand although is there should not even be a need to `prisma generate` in the first place because the DB schema [is already generated for the non_root container](https://github.com/BerriAI/litellm/blob/01c6cbd2705c65119b46866c0b4a44c9d78c662f/docker/Dockerfile.non_root#L87)

@krrishdholakia Would you accept a solution introducing a new environment variable which is set only in th e non_root image by incorporating it into the Dockerfile?
The migration script could then check for if this variable is set and skip the schema generation if so.
If sounds like a good solution, i would be happy to provide a PR.

**Comment by ghost** [comment]
Hey @freinold we actually just added prisma migrate deploy support - https://github.com/BerriAI/litellm/pull/9637

This should eliminate the schema generation step entirely - 

![Image](https://github.com/user-attachments/assets/9848c4b0-9cf1-4dcb-affb-eaf894824b07)

Can you try v1.61.14-stable (earliest version with the flag) with the cli arg `--use_prisma_migrate`  - https://github.com/BerriAI/litellm/blob/31082344a48b7e140d895dc0799f49571f715bf7/litellm/proxy/proxy_cli.py#L455

and let me know if that fixes it for you?

**Comment by freinold** [comment]
> Hey [@freinold](https://github.com/freinold) we actually just added prisma migrate deploy support - [#9637](https://github.com/BerriAI/litellm/pull/9637)
> 
> This should eliminate the schema generation step entirely -
> 
> ![Image](https://github.com/user-attachments/assets/9848c4b0-9cf1-4dcb-affb-eaf894824b07)
> 
> Can you try v1.61.14-stable (earliest version with the flag) with the cli arg `--use_prisma_migrate` -
> 
> [litellm/litellm/proxy/proxy_cli.py](https://github.com/BerriAI/litellm/blob/31082344a48b7e140d895dc0799f49571f715bf7/litellm/proxy/proxy_cli.py#L455)
> 
> Line 455 in [3108234](/BerriAI/litellm/commit/31082344a48b7e140d895dc0799f49571f715bf7)
>      "--use_prisma_migrate", 
> 
> and let me know if that fixes it for you?

Will try soon, thank you!

**Comment by mohittalele** [comment]
@krrishdholakia Does that mean there are two ways to do schema migration ? 

One way as per docs is :- https://docs.litellm.ai/docs/proxy/prod#7-use-helm-presync-hook-for-database-migrations-beta   
Using `DISABLE_SCHEMA_UPDATE=false`  and putting `DISABLE_SCHEMA_UPDATE=True` on migration job pod. This does not work with version litellm-nonroot image `main-v1.65.4-stable` 

Other way is using `USE_PRISMA_MIGRATE="True"`. I believe then migration job with image 1.65.4-stable and env `USE_PRISMA_MIGRATE="true"` variable will fix issue. In this case, How do you tell deployment pods to not do migration ? 
 Does `DISABLE_SCHEMA_UPDATE=true`  is no longer needed for deployment pods ?

**Comment by freinold** [comment]
> Hey [@freinold](https://github.com/freinold) we actually just added prisma migrate deploy support - [#9637](https://github.com/BerriAI/litellm/pull/9637)
> 
> This should eliminate the schema generation step entirely -
> 
> ![Image](https://github.com/user-attachments/assets/9848c4b0-9cf1-4dcb-affb-eaf894824b07)
> 
> Can you try v1.61.14-stable (earliest version with the flag) with the cli arg `--use_prisma_migrate` -
> 
> [litellm/litellm/proxy/proxy_cli.py](https://github.com/BerriAI/litellm/blob/31082344a48b7e140d895dc0799f49571f715bf7/litellm/proxy/proxy_cli.py#L455)
> 
> Line 455 in [3108234](/BerriAI/litellm/commit/31082344a48b7e140d895dc0799f49571f715bf7)
>      "--use_prisma_migrate", 
> 
> and let me know if that fixes it for you?

@krrishdholakia we finally had some time to try this fix today, it now works on a clean DB.
Migrating the running services will be fun but nothing to hard i hope.

But like @mohittalele I'm a bit puzzled on what to configure for the future:
Do we still need the migration job from the helm chart or did this become unneccesary now?
If I'm correct the migration is now run inside the application at startup, not in the dedicated job.

**Comment by ghost** [comment]
Hey @freinold @mohittalele, good point, we don't yet support prisma migrate on the helm step before starting applications 

We should do that, which would then replace the need for DISABLE_SCHEMA_UPDATE (since prisma migrate deploy implicitly does that)

**Comment by crackcomm** [comment]
I couldn't get it to work on nix, what I'm puzzled by is why it's trying to write the schema to `site-packages/prisma/schema.prisma`?

From what I gathered, it's because there is no `output` in `generator client { … } `.

**Comment by krrishdholakia** [comment]
Does prisma migrate deploy with custom migration dir solve this? - https://docs.litellm.ai/docs/proxy/prod#9-use-prisma-migrate-deploy

https://docs.litellm.ai/docs/proxy/prod#read-only-file-system

**Comment by vvidovic** [comment]
> Does prisma migrate deploy with custom migration dir solve this? - https://docs.litellm.ai/docs/proxy/prod#9-use-prisma-migrate-deploy
> 
> https://docs.litellm.ai/docs/proxy/prod#read-only-file-system

In the end, the LiteLLM migration container successfully completed and the LiteLLM container successfully started.

However, the log from the (completed) migration job still showed the error:
```
07:01:27 - LiteLLM Proxy:ERROR: prisma_migration.py:77 - 'prisma generate' stderr: Error: 
Traceback (most recent call last):
  File "/usr/local/lib/python3.13/site-packages/prisma/generator/generator.py", line 112, in run
    self._on_request(request)
    ~~~~~~~~~~~~~~~~^^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/prisma/generator/generator.py", line 170, in _on_request
    self.generate(data)
    ~~~~~~~~~~~~~^^^^^^
  File "/usr/local/lib/python3.13/site-packages/prisma/generator/generator.py", line 255, in generate
    shutil.copy(data.schema_path, packaged_schema)
    ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/shutil.py", line 428, in copy
    copyfile(src, dst, follow_symlinks=follow_symlinks)
    ~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/shutil.py", line 262, in copyfile
    with open(dst, 'wb') as fdst:
         ~~~~^^^^^^^^^^^
OSError: [Errno 30] Read-only file system: '/usr/local/lib/python3.13/site-packages/prisma/schema.prisma'
```

I tested the Helm upgrade by switching between a previous version and the new version, using the following env variables on both the migrationJob and the main pod, and I am not sure which of those helped in the end:
```
    # LITELLM_MIGRATION_DIR: "/nonexistent"
    # PRISMA_BINARY_CACHE_DIR: "/nonexistent"
    # DISABLE_SCHEMA_UPDATE: True
    # USE_PRISMA_MIGRATE: True
```

**Comment by SebLz** [comment]
Hello, I've just registered a new bug that I think is highly related and gives some more infos on how to solve this: https://github.com/BerriAI/litellm/issues/13208
